File: vmtools/src/main/java/com/vmloft/develop/library/tools/widget/VMLineView.java
package com.vmloft.develop.library.tools.widget;

import android.content.Context;
import android.content.res.TypedArray;
import android.os.Build;
import android.support.annotation.DrawableRes;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.vmloft.develop.library.tools.R;
import com.vmloft.develop.library.tools.utils.VMStr;

/**
 * Create by lzan13 on 2019/5/12 22:25
 *
 * 自定义单行控件，主要用于设置选项
 */
public class VMLineView extends RelativeLayout {

    // 左侧图标
    private ImageView mIconView;
    // 标题
    private TextView mTitleView;
    // 说明图标
    private ImageView mCaptionIconView;
    // 说明文本
    private TextView mCaptionView;
    // 右侧图标
    private ImageView mRightIconView;
    // 底部描述文本
    private TextView mDescriptionView;
    // 分割线
    private View mDecorationView;

    private int mIconRes;
    private String mTitle;
    private int mTitleColor;
    private int mCaptionIconRes;
    private String mCaption;
    private int mRightIconRes;
    private String mDescription;
    private boolean mDecoration;

    public VMLineView(Context context) {
        this(context, null);
    }

    public VMLineView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public VMLineView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs);
    }

    /**
     * 初始化
     */
    private void init(Context context, AttributeSet attrs) {
        LayoutInflater.from(context).inflate(R.layout.vm_widget_line_view, this);

        mIconView = findViewById(R.id.vm_line_icon_iv);
        mTitleView = findViewById(R.id.vm_line_title_tv);
        mCaptionIconView = findViewById(R.id.vm_line_caption_icon_iv);
        mCaptionView = findViewById(R.id.vm_line_caption_tv);
        mRightIconView = findViewById(R.id.vm_line_right_icon_iv);
        mDescriptionView = findViewById(R.id.vm_line_description_tv);
        mDecorationView = findViewById(R.id.vm_line_decoration);

        // 获取控件的属性值
        handleAttrs(context, attrs);

        setClickable(true);

        setupView();
    }

    /**
     * 获取资源属性
     */
    private void handleAttrs(Context context, AttributeSet attrs) {
        // 获取控件的属性值
        if (attrs == null) {
            return;
        }

        TypedArray array = context.obtainStyledAttributes(attrs, R.styleable.VMLineView);
        mIconRes = array.getResourceId(R.styleable.VMLineView_vm_line_icon, mIconRes);
        mTitle = array.getString(R.styleable.VMLineView_vm_line_title);
        mTitleColor = array.getColor(R.styleable.VMLineView_vm_line_title_color, mTitleColor);
        mCaptionIconRes = array.getResourceId(R.styleable.VMLineView_vm_line_caption_icon, mCaptionIconRes);
        mCaption = array.getString(R.styleable.VMLineView_vm_line_caption);
        mRightIconRes = array.getResourceId(R.styleable.VMLineView_vm_line_right_icon, mRightIconRes);
        mDescription = array.getString(R.styleable.VMLineView_vm_line_description);
        mDecoration = array.getBoolean(R.styleable.VMLineView_vm_line_decoration, mDecoration);
        array.recycle();
    }

    /**
     * 装载控件内容
     */
    private void setupView() {
        if (mIconRes == 0) {
            mIconView.setVisibility(GONE);
        } else {
            mIconView.setVisibility(VISIBLE);
            mIconView.setImageResource(mIconRes);
        }
        if (!VMStr.isEmpty(mTitle)) {
            mTitleView.setText(mTitle);
        }
        if (mTitleColor != 0) {
            mTitleView.setTextColor(mTitleColor);
        }

        if (mCaptionIconRes == 0) {
            mCaptionIconView.setVisibility(GONE);
        } else {
            mCaptionIconView.setVisibility(VISIBLE);
            mCaptionIconView.setImageResource(mCaptionIconRes);
        }
        if (VMStr.isEmpty(mCaption)) {
            mCaptionView.setVisibility(GONE);
        } else {
            mCaptionView.setVisibility(VISIBLE);
            mCaptionView.setText(mCaption);
        }
        if (mRightIconRes == 0) {
            mRightIconView.setVisibility(GONE);
        } else {
            mRightIconView.setVisibility(VISIBLE);
            mRightIconView.setImageResource(mRightIconRes);
        }
        if (VMStr.isEmpty(mDescription)) {
            mDescriptionView.setVisibility(GONE);
        } else {
            mDescriptionView.setVisibility(VISIBLE);
            mDescriptionView.setText(mDescription);
        }
        if (mDecoration) {
            mDecorationView.setVisibility(VISIBLE);
        } else {
            mDecorationView.setVisibility(GONE);
        }
    }

    /**
     * ----------------------------------- 内容设置 -----------------------------------
     *
     * 设置图标
     */
    public void setIconRes(@DrawableRes int resId) {
        mIconRes = resId;
        if (mIconRes == 0) {
            mIconView.setVisibility(GONE);
            return;
        }
        mIconView.setVisibility(VISIBLE);
        mIconView.setImageResource(mIconRes);
    }

    /**
     * 设置标题
     */
    public void setTitle(String title) {
        mTitle = title;
        mTitleView.setText(mTitle);
    }

    /**
     * 设置标题颜色
     */
    public void setTitleColor(int color) {
        mTitleColor = color;
        mTitleView.setTextColor(mTitleColor);
    }

    /**
     * 设置标题样式
     */
    public void setTitleStyle(int resId) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            mTitleView.setTextAppearance(resId);
        } else {
            mTitleView.setTextAppearance(getContext(), resId);
        }
    }

    /**
     * 设置右侧描述图标
     */
    public void setCaptionIcon(@DrawableRes int resId) {
        mCaptionIconRes = resId;
        if (mCaptionIconRes == 0) {
            mCaptionIconView.setVisibility(GONE);
            return;
        }
        mCaptionIconView.setVisibility(VISIBLE);
        mCaptionIconView.setImageResource(mCaptionIconRes);
    }

    /**
     * 设置右侧描述内容
     */
    public void setCaption(String caption) {
        mCaption = caption;
        if (VMStr.isEmpty(mCaption)) {
            mCaptionView.setVisibility(GONE);
            return;
        }
        mCaptionView.setVisibility(VISIBLE);
        mCaptionView.setText(mCaption);
    }

    /**
     * 设置右侧描述样式
     */
    public void setCaptionStyle(int resId) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            mCaptionView.setTextAppearance(resId);
        } else {
            mCaptionView.setTextAppearance(getContext(), resId);
        }
    }

    /**
     * 设置右侧图标
     */
    public void setRightIcon(@DrawableRes int resId) {
        mRightIconRes = resId;
        if (mRightIconRes == 0) {
            mRightIconView.setVisibility(GONE);
            return;
        }
        mRightIconView.setVisibility(VISIBLE);
        mRightIconView.setImageResource(mRightIconRes);
    }

    /**
     * 设置底部描述
     */
    public void setDescription(String description) {
        mDescription = description;
        if (VMStr.isEmpty(mDescription)) {
            mDescriptionView.setVisibility(GONE);
            return;
        }
        mDescriptionView.setVisibility(VISIBLE);
        mDescriptionView.setText(mDescription);
    }

    /**
     * 设置激活状态
     */
    @Override
    public void setActivated(boolean activated) {
        super.setActivated(activated);
        mRightIconView.setActivated(activated);
    }
}

File: vmtools/src/main/java/com/vmloft/develop/library/tools/widget/VMLineView.kt
package com.vmloft.develop.library.tools.widget

import android.content.Context
import android.os.Build.VERSION
import android.os.Build.VERSION_CODES
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.RelativeLayout
import android.widget.TextView

import com.vmloft.develop.library.tools.R
import com.vmloft.develop.library.tools.R.layout
import com.vmloft.develop.library.tools.R.styleable
import com.vmloft.develop.library.tools.utils.VMColor
import com.vmloft.develop.library.tools.utils.VMDimen
import com.vmloft.develop.library.tools.utils.logger.VMLog

/**
 * Create by lzan13 on 2019/5/12 22:25
 *
 * 自定义单行控件，主要用于设置选项
 */
class VMLineView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : RelativeLayout(context, attrs, defStyleAttr) {

    private var vmLineRootView: View // 根布局
    private var mIconView: ImageView // 左侧图标
    private var mTitleView: TextView // 标题
    private var mRightContainer: LinearLayout // 右侧容器
    private var mCaptionIconView: ImageView // 说明图标
    private var mCaptionView: TextView // 说明文本
    private var mRightIconView: ImageView // 右侧图标
    private var mDescriptionView: TextView // 底部描述文本
    private var mBottomContainer: LinearLayout // 底部容器
    private var mDecorationView: View // 分割线

    // 分割线辅助定位
    private var mDecorationAidedStartView: View
    private var mDecorationAidedEndView: View

    private var mIconRes = 0 // 图标
    private var mIconSize = 24 // 图标大小

    private var mTitle: String? = null
    private var mTitleColor = 0
    private var mTitleStyle = 0
    private var mTitleSpace = 0 // 标题内容边距

    private var mCaptionIconRes = 0
    private var mCaption: String? = null
    private var mCaptionColor = 0
    private var mCaptionStyle = 0
    private var mCaptionSpace = 0 // 右侧描述内容边距

    private var mRightIconRes = 0 // 右侧图标
    private var mRightIconSize = 24 // 右侧图标大小

    private var mDescription: String? = null
    private var mDescriptionColor = 0
    private var mDescriptionStyle = 0 // 底部描述样式

    private var mDecoration = 0 // 是否展示分割线
    private var mDecorationHeight = 0 // 分割线高度
    private var mDecorationColor = 0 // 分割线颜色
    private var mDecorationAidedStart = 0 // 分割线开始空间
    private var mDecorationAidedEnd = 0 // 分割线结束空间

    /**
     * 初始化
     */
    init {
        LayoutInflater.from(context).inflate(layout.vm_widget_line_view, this)

        mTitleColor = VMColor.byRes(R.color.vm_title)
        mCaptionColor = VMColor.byRes(R.color.vm_caption)
        mDescriptionColor = VMColor.byRes(R.color.vm_description)

        mTitleSpace = VMDimen.dp2px(16)
        mCaptionSpace = VMDimen.dp2px(8)

        mIconSize = VMDimen.dp2px(24)
        mRightIconSize = VMDimen.dp2px(24)

        mDecorationAidedStart = VMDimen.dp2px(16)
        mDecorationAidedEnd = VMDimen.dp2px(16)

        mDecorationHeight = VMDimen.getDimenPixel(R.dimen.vm_dimen_0_1)
        mDecorationColor = VMColor.byRes(R.color.vm_decoration)

        vmLineRootView = findViewById(R.id.vmLineRootView)
        mIconView = findViewById(R.id.vmLineIconIV)
        mTitleView = findViewById(R.id.vmLineTitleTV)
        mRightContainer = findViewById(R.id.vmLineRightContainer)
        mCaptionIconView = findViewById(R.id.vmLineCaptionIconIV)
        mCaptionView = findViewById(R.id.vmLineCaptionTV)
        mRightIconView = findViewById(R.id.vmLineRightIconIV)
        mDescriptionView = findViewById(R.id.vmLineDescriptionTV)
        mBottomContainer = findViewById(R.id.vmLineBottomContainer)
        mDecorationView = findViewById(R.id.vmLineDecoration)
        mDecorationAidedStartView = findViewById(R.id.vmLineDecorationAidedStartView)
        mDecorationAidedEndView = findViewById(R.id.vmLineDecorationAidedEndView)

        // 获取控件的属性值
        handleAttrs(context, attrs)
        isClickable = true
        setupView()
    }

    /**
     * 获取资源属性
     */
    private fun handleAttrs(context: Context, attrs: AttributeSet?) {
        // 获取控件的属性值
        if (attrs == null) {
            return
        }
        val array = context.obtainStyledAttributes(attrs, styleable.VMLineView)

        mIconRes = array.getResourceId(styleable.VMLineView_vm_line_icon, mIconRes)
        mIconSize = array.getDimensionPixelOffset(styleable.VMLineView_vm_line_icon_size, mIconSize)
        mTitle = array.getString(styleable.VMLineView_vm_line_title)
        mTitleColor = array.getColor(styleable.VMLineView_vm_line_title_color, mTitleColor)
        mTitleStyle = array.getResourceId(styleable.VMLineView_vm_line_title_style, mTitleStyle)
        mTitleSpace = array.getDimensionPixelOffset(styleable.VMLineView_vm_line_title_space, mTitleSpace)

        mCaptionIconRes = array.getResourceId(styleable.VMLineView_vm_line_caption_icon, mCaptionIconRes)
        mCaption = array.getString(styleable.VMLineView_vm_line_caption)
        mCaptionColor = array.getColor(styleable.VMLineView_vm_line_caption_color, mCaptionColor)
        mCaptionStyle = array.getResourceId(styleable.VMLineView_vm_line_caption_style, mCaptionStyle)
        mCaptionSpace = array.getDimensionPixelOffset(styleable.VMLineView_vm_line_caption_space, mCaptionSpace)

        mRightIconRes = array.getResourceId(styleable.VMLineView_vm_line_right_icon, mRightIconRes)
        mRightIconSize = array.getDimensionPixelOffset(styleable.VMLineView_vm_line_right_icon_size, mRightIconSize)

        mDescription = array.getString(styleable.VMLineView_vm_line_description)
        mDescriptionColor = array.getColor(styleable.VMLineView_vm_line_description_color, mDescriptionColor)
        mDescriptionStyle = array.getResourceId(styleable.VMLineView_vm_line_description_style, mDescriptionStyle)

        mDecoration = array.getInt(styleable.VMLineView_vm_line_decoration, mDecoration)
        mDecorationColor = array.getColor(styleable.VMLineView_vm_line_decoration_color, mDecorationColor)
        mDecorationAidedStart = array.getDimensionPixelOffset(styleable.VMLineView_vm_line_decoration_start, mDecorationAidedStart)
        mDecorationAidedEnd = array.getDimensionPixelOffset(styleable.VMLineView_vm_line_decoration_end, mDecorationAidedEnd)
        mDecorationHeight = array.getDimensionPixelOffset(styleable.VMLineView_vm_line_decoration_height, mDecorationHeight)
        array.recycle()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val height = measuredHeight
        val rootHeight = vmLineRootView.measuredHeight
        // 根据测量结果，将内部布局高度设置为与外层布局一样
        if (height > rootHeight) {
            vmLineRootView.layoutParams.height = height
        }
    }

    /**
     * 装载控件内容
     */
    private fun setupView() {
        setIconRes(mIconRes)
        setIconSize(mIconSize)

        setTitle(mTitle)
        setTitleColor(mTitleColor)
        setTitleStyle(mTitleStyle)

        setCaptionIcon(mCaptionIconRes)
        setCaption(mCaption)
        setCaptionColor(mCaptionColor)
        setCaptionStyle(mCaptionStyle)

        setRightIcon(mRightIconRes)
        setRightIconSize(mRightIconSize)

        setDescription(mDescription)
        setDescriptionColor(mDescriptionColor)
        setDescriptionStyle(mDescriptionStyle)

        if (mDecoration == 0) {
            mDecorationView.visibility = View.GONE
        } else {
            mDecorationView.visibility = View.VISIBLE
            mDecorationView.layoutParams.height = mDecorationHeight
            mDecorationView.setBackgroundColor(mDecorationColor)

            mDecorationAidedStartView.visibility = if (mDecoration == 1) View.GONE else View.VISIBLE
            mDecorationAidedEndView.visibility = if (mDecoration == 2) View.VISIBLE else View.GONE

            mDecorationAidedStartView.layoutParams.width = mDecorationAidedStart
            mDecorationAidedEndView.layoutParams.width = mDecorationAidedEnd
        }
    }

    /**
     * ----------------------------------- 内容设置 -----------------------------------
     *
     * 设置图标
     */
    fun setIconRes(resId: Int) {
        mIconRes = resId
        if (mIconRes == 0) {
            mIconView.visibility = View.GONE
        } else {
            mIconView.visibility = View.VISIBLE
            mIconView.setImageResource(mIconRes)
        }
    }

    /**
     * 设置图标大小
     */
    fun setIconSize(size: Int) {
        mIconSize = size
        mIconView.layoutParams.width = mIconSize
        mIconView.layoutParams.height = mIconSize
    }

    /**
     * 设置标题
     */
    fun setTitle(title: String?) {
        mTitle = title
        mTitleView.text = if (mTitle.isNullOrEmpty()) "" else mTitle
        mTitleView.setPadding(mTitleSpace, 0, mTitleView.paddingEnd, 0)
    }

    /**
     * 设置标题颜色
     */
    fun setTitleColor(color: Int) {
        if (color == 0) return
        mTitleColor = color
        mTitleView.setTextColor(mTitleColor)
    }

    /**
     * 设置标题样式
     */
    fun setTitleStyle(resId: Int) {
        if (mTitleStyle == 0) return
        mTitleStyle = resId
        if (VERSION.SDK_INT >= VERSION_CODES.M) {
            mTitleView.setTextAppearance(mTitleStyle)
        } else {
            mTitleView.setTextAppearance(context, mTitleStyle)
        }
    }

    /**
     * 添加控件到右侧容器
     */
    fun setRightView(view: View?) {
        if (view != null) {
            mRightContainer.removeAllViews()
            mRightContainer.addView(view)
        } else {
            mRightContainer.removeAllViews()
        }
    }

    /**
     * 设置右侧描述图标
     */
    fun setCaptionIcon(resId: Int) {
        mCaptionIconRes = resId
        if (mCaptionIconRes == 0) {
            mCaptionIconView.visibility = View.GONE
        } else {
            mCaptionIconView.visibility = View.VISIBLE
            mCaptionIconView.setImageResource(mCaptionIconRes)
        }
    }

    /**
     * 设置右侧说明文本内容
     */
    fun setCaption(caption: String?) {
        mCaption = caption
        if (mCaption.isNullOrEmpty()) {
            mCaptionView.visibility = View.GONE
        } else {
            mCaptionView.visibility = View.VISIBLE
            mCaptionView.text = mCaption
            mCaptionView.setPadding(0, 0, mCaptionSpace, 0)
        }
    }

    /**
     * 设置右侧说明文本颜色
     */
    fun setCaptionColor(color: Int) {
        if (color == 0) return
        mCaptionColor = color
        mCaptionView.setTextColor(mCaptionColor)
    }

    /**
     * 设置右侧说明文本样式
     */
    fun setCaptionStyle(resId: Int) {
        if (mCaptionStyle == 0) return
        mCaptionStyle = resId
        if (VERSION.SDK_INT >= VERSION_CODES.M) {
            mCaptionView.setTextAppearance(mCaptionStyle)
        } else {
            mCaptionView.setTextAppearance(context, mCaptionStyle)
        }
    }

    /**
     * 设置右侧图标
     */
    fun setRightIcon(resId: Int) {
        mRightIconRes = resId
        if (mRightIconRes == 0) {
            mRightIconView.visibility = View.GONE
        } else {
            mRightIconView.visibility = View.VISIBLE
            mRightIconView.setImageResource(mRightIconRes)
        }
    }

    /**
     * 设置右侧图标大小
     */
    fun setRightIconSize(size: Int) {
        mRightIconSize = size
        mRightIconView.layoutParams.width = mRightIconSize
        mRightIconView.layoutParams.height = mRightIconSize
    }

    /**
     * 设置底部描述
     */
    fun setDescription(description: String?) {
        mDescription = description
        if (mDescription.isNullOrEmpty()) {
            mDescriptionView.visibility = View.GONE
        } else {
            mDescriptionView.visibility = View.VISIBLE
            mDescriptionView.text = mDescription
            mDescriptionView.setPadding(mTitleSpace, 0, mDescriptionView.paddingEnd, 0)
        }
    }

    /**
     * 设置右侧说明文本颜色
     */
    fun setDescriptionColor(color: Int) {
        if (color == 0) return
        mDescriptionColor = color
        mDescriptionView.setTextColor(mDescriptionColor)
    }

    /**
     * 设置右侧说明文本样式
     */
    fun setDescriptionStyle(resId: Int) {
        if (mDescriptionStyle == 0) return
        mDescriptionStyle = resId
        if (VERSION.SDK_INT >= VERSION_CODES.M) {
            mDescriptionView.setTextAppearance(mDescriptionStyle)
        } else {
            mDescriptionView.setTextAppearance(context, mDescriptionStyle)
        }
    }

    /**
     * 添加控件到底部容器
     */
    fun setBottomView(view: View?) {
        if (view != null) {
            mBottomContainer.removeAllViews()
            mBottomContainer.addView(view)
        } else {
            mBottomContainer.removeAllViews()
        }
    }

    /**
     * 设置激活状态
     */
    override fun setActivated(activated: Boolean) {
        super.setActivated(activated)
        mRightIconView.isActivated = activated
    }

}